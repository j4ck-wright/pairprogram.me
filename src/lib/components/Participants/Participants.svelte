<script lang="ts">
	import { setParticipants, watchParticipants, participantsStore } from '$lib/firebase';
	import { ParticipantBadge } from '../';

	let newParticipant: string;
	let active = false;

	function removeParticipant(event: { detail: { participant: string } }) {
		const removedParticipant = event.detail.participant;
		const index = $participantsStore.indexOf(removedParticipant);
		$participantsStore.splice(index, 1);
		setParticipants($participantsStore);
	}

	function toggleActive() {
		active = !active;
	}

	watchParticipants();
</script>

<div class="container flex items-start gap-2 w-4/5">
	<span class="">Participants:</span>
	<div class="badges flex flex-wrap gap-2" data-testid="participantBadges">
		{#each $participantsStore as participant}
			<ParticipantBadge {participant} on:removeParticipant={removeParticipant} />
		{/each}
		<div
			class="bg-primary rounded-full outline outline-primary bg-transparent flex items-center"
			data-testid="addParticipantBtn"
		>
			<iconify-icon
				role="button"
				tabindex="0"
				icon={active ? 'mdi:pencil' : 'ic:baseline-plus'}
				on:click={toggleActive}
				on:keydown={toggleActive}
				class="cursor-pointer text-primary {active && 'ml-1'} text-[24px]"
			/>
			{#if active}
				<!-- svelte-ignore a11y-autofocus -->
				<input
					autofocus
					class="mr-2 outline-none bg-transparent"
					data-testid="participantInput"
					type="text"
					size="3"
					bind:value={newParticipant}
					on:keydown={(e) => {
						if (e.key === 'Enter' && newParticipant && newParticipant.length > 0) {
							setParticipants([...$participantsStore, newParticipant]);
							newParticipant = '';
						}
					}}
					on:focusout={() => {
						active = false;
						newParticipant = '';
					}}
				/>
			{/if}
		</div>
	</div>
</div>
