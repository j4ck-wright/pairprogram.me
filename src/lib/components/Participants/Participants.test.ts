import { render, screen } from '@testing-library/svelte';
import Participants from './Participants.svelte';

describe('Participants test suite', () => {
	beforeEach(() => {
		render(Participants);
	});

	it('Should render zero participants on page load', () => {
		const participantBadges = screen.getByTestId('participantBadges');
		expect(participantBadges).toBeTruthy();
		expect(participantBadges.childElementCount).toBe(1);
	});

	it('Should display the add participant button', () => {
		const addParticipantBtn = screen.getByTestId('addParticipantBtn');
		expect(addParticipantBtn).toBeTruthy();
	});
});

describe('Participants snapshot', () => {
	it('Should match snapshot', () => {
		const participantComponent = render(Participants);
		expect(participantComponent).toMatchSnapshot();
	});
});
