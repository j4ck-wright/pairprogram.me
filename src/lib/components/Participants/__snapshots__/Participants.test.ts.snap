// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Participants snapshot > Should match snapshot 1`] = `
{
  "component": Participants {
    "$$": {
      "after_update": [],
      "before_update": [],
      "bound": {},
      "callbacks": {},
      "context": Map {},
      "ctx": [
        [],
        undefined,
        false,
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": [
        -1,
      ],
      "fragment": {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": [
        [Function],
      ],
      "on_disconnect": [],
      "on_mount": [],
      "props": {},
      "root": <div>
        <div
          class="container flex items-start gap-2 w-4/5"
        >
          <span
            class=""
          >
            Participants:
          </span>
           
          <div
            class="badges flex flex-wrap gap-2"
            data-testid="participantBadges"
          >
             
            <div
              class="bg-primary rounded-full outline outline-primary bg-transparent flex items-center"
              data-testid="addParticipantBtn"
            >
              <iconify-icon
                class="cursor-pointer text-primary false text-[24px]"
                icon="ic:baseline-plus"
                role="button"
                tabindex="0"
              />
               
            </div>
          </div>
        </div>
      </div>,
      "skip_bound": false,
      "update": [Function],
    },
    "$$events_def": undefined,
    "$$prop_def": undefined,
    "$$set": undefined,
    "$$slot_def": undefined,
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="container flex items-start gap-2 w-4/5"
      >
        <span
          class=""
        >
          Participants:
        </span>
         
        <div
          class="badges flex flex-wrap gap-2"
          data-testid="participantBadges"
        >
           
          <div
            class="bg-primary rounded-full outline outline-primary bg-transparent flex items-center"
            data-testid="addParticipantBtn"
          >
            <iconify-icon
              class="cursor-pointer text-primary false text-[24px]"
              icon="ic:baseline-plus"
              role="button"
              tabindex="0"
            />
             
          </div>
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
